packages/runtime/src/widgets/Search.tsx

```diff
@@ -18,51 +24,108 @@ export type SearchProps = {
   placeholder?: string;
   data?: Expression<object>;
   searchKey?: string;
+  onValueChange?: EnsembleAction;
+  onSelect?: EnsembleAction;
 } & EnsembleWidgetProps<SearchStyles>;
 
 export const Search: React.FC<SearchProps> = ({
+  id,
   placeholder,
   data,
   searchKey,
+  onValueChange,
+  onSelect,
   styles,
 }) => {
   const [options, setOptions] = useState<SelectProps<object>["options"]>([]);
+  const [value, setValue] = useState("");
+  const [filteredOptions, setFilteredOptions] = useState<
+    SelectProps<object>["options"]
+  >([]);
+  const templateData = useTemplateData(data ?? "");
+  const valueChangeAction = useEnsembleAction(onValueChange);
+  const selectAction = useEnsembleAction(onSelect);
 
-  const templateData = useTemplateData(data!);
+  useRegisterBindings(
+    {
+      value,
+      options,
+    },
+    id,
+    { setValue, setOptions },
+  );
+
+  const handleSearch = (search: string): void => {
+    if (!searchKey) return;
 
-  // TODO: Pass in search predicate function via props or filter via API
-  const handleSearch = (value: string) => {
-    if (Array.isArray(templateData)) {
-      setOptions(
-        value
+    if (onValueChange?.invokeApi) {
+      (async (): Promise<void> => {
+        const apiResponse = await valueChangeAction?.callback();
+
+        if (apiResponse)
+          setFilteredOptions(
+            findValuesByKey<string>(apiResponse, searchKey)
+              .filter((item) =>
+                item.toLowerCase().includes(search.toLowerCase()),
+              )
+              .map((item) => ({
+                value: item,
+                label: item,
+              })),
+          );
```

Comment:
The idea here is if onValueChange invokesApi, the API will return the filtered values and you just provide the data binding, e.g.

```
Search:
  options: mySearchEndpoint.body.results
  onValueChange:
     invokeApi
        name: mySearchEndpoint
        inputs:
           searchKey: value
```

Reply:
So we don't need to do this in react code just invoke the callback if it's there

Reply:
got it